examples/energy-management-app/energy-management-common/src/EVSEManufacturerImpl.cpp

```diff
@@ -93,68 +94,114 @@ CHIP_ERROR EVSEManufacturer::Shutdown()
+
 /**
  * @brief   Allows a client application to send in power readings into the system
  *
- * @param[in]  aEndpointId     - Endpoint to send to EPM Cluster
- * @param[in]  aActivePower_mW - Power measured in milli-watts
- * @param[in]  aVoltage_mV     - Voltage measured in milli-volts
- * @param[in]  aCurrent_mA     - Current measured in milli-amps
+ * @param[in]  aEndpointId       - Endpoint to send to EPM Cluster
+ * @param[in]  aActivePower_mW   - ActivePower measured in milli-watts
+ * @param[in]  aVoltage_mV       - Voltage measured in milli-volts
+ * @param[in]  aActiveCurrent_mA - ActiveCurrent measured in milli-amps
  */
 CHIP_ERROR EVSEManufacturer::SendPowerReading(EndpointId aEndpointId, int64_t aActivePower_mW, int64_t aVoltage_mV,
-                                              int64_t aCurrent_mA)
+                                              int64_t aActiveCurrent_mA)
 {
-    // TODO add Power Readings when EPM cluster is merged
+    EVSEManufacturer * mn = GetEvseManufacturer();
+    VerifyOrReturnError(mn != nullptr, CHIP_ERROR_UNINITIALIZED);
+
+    ElectricalPowerMeasurementDelegate * dg = mn->GetEPMDelegate();
+    VerifyOrReturnError(dg != nullptr, CHIP_ERROR_UNINITIALIZED);
+
+    dg->SetActivePower(MakeNullable(aActivePower_mW));
+    dg->SetVoltage(MakeNullable(aVoltage_mV));
+    dg->SetActiveCurrent(MakeNullable(aActiveCurrent_mA));
 
     return CHIP_NO_ERROR;
 }
 
 /**
- * @brief   Allows a client application to send in energy readings into the system
+ * @brief   Allows a client application to send cumulative energy readings into the system
  *
  *          This is a helper function to add timestamps to the readings
  *
  * @param[in]  aCumulativeEnergyImported -total energy imported in milli-watthours
  * @param[in]  aCumulativeEnergyExported -total energy exported in milli-watthours
  */
-CHIP_ERROR EVSEManufacturer::SendEnergyReading(EndpointId aEndpointId, int64_t aCumulativeEnergyImported,
-                                               int64_t aCumulativeEnergyExported)
+CHIP_ERROR EVSEManufacturer::SendCumulativeEnergyReading(EndpointId aEndpointId, int64_t aCumulativeEnergyImported,
+                                                         int64_t aCumulativeEnergyExported)
 {
     MeasurementData * data = MeasurementDataForEndpoint(aEndpointId);
 
     EnergyMeasurementStruct::Type energyImported;
     EnergyMeasurementStruct::Type energyExported;
 
     // Get current timestamp
-    uint32_t currentTimestamp;
+    uint32_t currentTimestamp = 0;
+
+    // In case we can't get real time
+    System::Clock::Milliseconds64 system_time_ms =
+        std::chrono::duration_cast<System::Clock::Milliseconds64>(chip::Server::GetInstance().TimeSinceInit());
+    uint64_t nowMS = static_cast<uint64_t>(system_time_ms.count());
+
     CHIP_ERROR err = GetEpochTS(currentTimestamp);
     if (err != CHIP_NO_ERROR)
     {
         ChipLogError(AppServer, "GetEpochTS returned error getting timestamp");
-        return err;
```

Comment:
Error is not forwarded anymore. In this case would it make sense to:
  - add the error as CHIP_ERROR_FORMAT to the log
  - set currentTimestamp to a specific value, since I GetEpochTS does not seem to specify "on error the input value is unchanged" ?
  
It seems you need to set currentTimestamp = nowMS and then you do not need the conditional below.

It also seems that you separated these timestamps readouts a bit from line 192 when actually used. How about moving these lower so that the initialization and usage are closer together? nowMS seems to only be needed on the `if error` branch

Reply:
Yep - made that change - much clearer thanks.